api/leaderworkerset/v1/leaderworkerset_types.go

```diff
@@ -89,12 +97,71 @@ type LeaderWorkerTemplate struct {
 	// Number of pods to create. It is the total number of pods in each group.
 	// The minimum is 1 which represent the leader. When set to 1, the leader
 	// pod is created for each group as well as a 0-replica StatefulSet for the workers.
-	Size *int32 `json:"size"`
+	// Default to 1.
+	// +optional
+	Size *int32 `json:"size,omitempty"`
 
 	// RestartPolicy defines the restart policy when pod failures happen.
 	RestartPolicy RestartPolicyType `json:"restartPolicy,omitempty"`
 }
 
+// RolloutStrategy defines the strategy that the leaderWorkerSet controller
+// will use to perform replica updates.
+type RolloutStrategy struct {
+	// Type defines the rollout strategy, it can only be “RollingUpdate” for now.
+	//
+	// +kubebuilder:validation:Enum={RollingUpdate}
+	// +kubebuilder:default="RollingUpdate"
+	Type RolloutStrategyType `json:"type,omitempty"`
+
+	// RollingUpdateConfiguration defines the parameters to be used when type is RollingUpdateStrategyType.
+	// +optional
+	RollingUpdateConfiguration *RollingUpdateConfiguration `json:"rollingUpdateConfiguration,omitempty"`
+}
+
+// RollingUpdateConfiguration defines the parameters to be used for RollingUpdateStrategyType.
+type RollingUpdateConfiguration struct {
+	// The maximum number of replica that can be unavailable during the update.
+	// Value can be an absolute number (ex: 5) or a percentage of total replicas at the start of update (ex: 10%).
+	// Absolute number is calculated from percentage by rounding down.
+	// This can not be 0 if MaxSurge is 0.
+	// By default, a fixed value of 1 is used.
+	// Example: when this is set to 30%, the old replicas can be scaled down by 30%
+	// immediately when the rolling update starts. Once new replicas are ready, old replicas
+	// can be scaled down further, followed by scaling up the new replicas, ensuring
+	// that at least 70% of original number of replicas are available at all times
+	// during the update.
+	//
+	// +kubebuilder:validation:XIntOrString
+	// +kubebuilder:default=1
+	MaxUnavailable intstr.IntOrString `json:"maxUnavailable,omitempty"`
+
+	// The maximum number of replicas that can be scheduled above the original number of
+	// replicas.
+	// Value can be an absolute number (ex: 5) or a percentage of total replicas at
+	// the start of the update (ex: 10%).
+	// This can not be 0 if MaxUnavailable is 0.
+	// Absolute number is calculated from percentage by rounding up.
+	// By default, a value of 1 is used.
```

Comment:
if we didn't specify MaxUnavailable and MaxSurge, the default value for them are both 1? which means that actually we could have 2 replicas upgrading at the same time?

Reply:
But I think it is reasonable to have both as 1 at the same time